ml-agents/mlagents/trainers/torch/utils.py

```diff
@@ -15,6 +15,132 @@
 from mlagents.trainers.torch.distributions import DistInstance, DiscreteDistInstance
 
 
+class AgentAction(NamedTuple):
+    continuous_tensor: torch.Tensor
+    discrete_list: List[torch.Tensor]
+
+    @property
+    def discrete_tensor(self):
+        return torch.stack(self.discrete_list, dim=-1)
+
+    def to_numpy_dict(self) -> Dict[str, np.ndarray]:
+        array_dict: Dict[str, np.ndarray] = {}
+        if self.continuous_tensor is not None:
+            array_dict["continuous_action"] = ModelUtils.to_numpy(
+                self.continuous_tensor
+            )
+        if self.discrete_list is not None:
+            array_dict["discrete_action"] = ModelUtils.to_numpy(
+                self.discrete_tensor[:, 0, :]
+            )
+        return array_dict
+
+    def to_tensor_list(self) -> List[torch.Tensor]:
+        tensor_list: List[torch.Tensor] = []
+        if self.continuous_tensor is not None:
+            tensor_list.append(self.continuous_tensor)
+        if self.discrete_list is not None:
+            tensor_list += (
+                self.discrete_list
+            )  # Note this is different for ActionLogProbs
+        return tensor_list
+
+    @staticmethod
+    def create(
+        tensor_list: List[torch.Tensor], action_spec: ActionSpec
+    ) -> "AgentAction":
+        continuous: torch.Tensor = None
+        discrete: List[torch.Tensor] = None
+        _offset = 0
+        if action_spec.continuous_size > 0:
+            continuous = tensor_list[0]
+            _offset = 1
+        if action_spec.discrete_size > 0:
+            discrete = tensor_list[_offset:]
+        return AgentAction(continuous, discrete)
+
+    @staticmethod
+    def extract(buff: Dict[str, np.ndarray]) -> "AgentAction":
+        continuous: torch.Tensor = None
+        discrete: List[torch.Tensor] = None
+        if "continuous_action" in buff:
+            continuous = ModelUtils.list_to_tensor(buff["continuous_action"])
+        if "discrete_action" in buff:
+            discrete_tensor = ModelUtils.list_to_tensor(buff["discrete_action"])
+            discrete = [
+                discrete_tensor[..., i] for i in range(discrete_tensor.shape[-1])
+            ]
+        return AgentAction(continuous, discrete)
+
+
+class ActionLogProbs(NamedTuple):
+    continuous_tensor: torch.Tensor
+    discrete_list: List[torch.Tensor]
+    all_discrete_list: List[torch.Tensor]
```

Comment:
what is `all_discrete_list` for?

Reply:
SAC uses all_log_probs for discrete actions instead of just the log probs of the actions taken.

Reply:
log_probs is the probability of a single action, all_log_probs is all the probs of actions on a given branch